libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,463 @@
+        // if we got here, we have identified a chain of overlapping loops,
+        // starting with loop at index i, ending at i+chain_len
+        // now, decide if it would be better to delete even or odd loops
+        float even_dist_removed = 0;
+        float odd_dist_removed = 0;
+        // for each prunable loop in the overlapping chain
+        uint32_t k;
+        for (k = i; k <= i + chain_len; k++) {
+            float loop_len = 0;
+            // for each point in said prunable loop
+            for (uint32_t l = _prunable_loops[k].start_index; l < _prunable_loops[k].end_index; l++) {
+                // add the real-world distance between two sequential points in the loop
+                loop_len += HYPOT(path[l-1], path[l]);
+            }
+            // now add the real-world length of that loop to the correct counter
+            if (k%2==0) { // evens
+                even_dist_removed += loop_len;
+            } else {   //odds
+                odd_dist_removed += loop_len;
+            }
+        }
+        // if we want to delete evens, but the loops starts at odds, or if we want to delete odds but the loop starts at evens, offset deletion by one.
+        bool delete_offset = (even_dist_removed < odd_dist_removed) == (k % 2);
+        for (uint32_t l = i + delete_offset; l <= i + chain_len; l++) {
+            _prunable_loops.erase(_prunable_loops.begin() + l);
+            chain_len--;
+        }
+        i = j; // next iteration, only start looking for the next chain at an index greater than the end of this chain
+    }
+}
+
+/**
+*   Only prunes loops until $points_to_delete points have been removed. It does not necessarily prune all loops.
+*/
+void SafeRTL_Path::_zero_points_by_loops(uint8_t points_to_delete)
+{
+    int removed_points = 0;
+    for (uint32_t i = 0; i < _prunable_loops.size(); i++) {
+        loop l = _prunable_loops[i];
+        _pruning_clean_until = MIN(_pruning_clean_until, l.start_index-1);
+        for (uint32_t j = l.start_index; j < l.end_index; j++) {
+            // zero this point if it wasn't already zeroed
+            if (path[j] != Vector3f(0.0f, 0.0f, 0.0f)) {
+                if (_logging_enabled) {
+                    DataFlash_Class::instance()->Log_Write_SRTL(DataFlash_Class::SRTL_POINT_PRUNE, path[j]);
+                }
+                path[j] = Vector3f(0.0f, 0.0f, 0.0f);
+            }
+        }
+        path[int((l.start_index+l.end_index)/2.0)] = l.halfway_point;
+        removed_points += l.end_index - l.start_index - 1;
+        if (removed_points > points_to_delete) {
+            return;
+        }
+    }
+}
+
+/**
+*  Removes all NULL points from the path, and shifts remaining items to correct position.
+*  The first item will not be removed.
+*/
+void SafeRTL_Path::_remove_empty_points()
+{
+    uint32_t i = 0;
+    uint32_t j = 0;
+    uint32_t removed = 0;
+    while (++i <= _last_index) { // never removes the first item. This should always be {0,0,0}
```

Comment:
that's what `_last_index` is because i'm using a standard c style array. There's alsready a check to make sure that `_last_index` does not go higher than the end of the array.